### PR TITLE
improve kubectl config handling and UX

### PR DESCRIPTION
**Pull request by jaypipes**
Previously to
https://github.com/aws-controllers-k8s/test-infra/pull/222, when running `make kind-test`, there was some helpful output that set up the kubectl config so that developers could easily connect to the KinD cluster housing test pods. In addition, prior to the mentioned PR222, when scripts would run commands in a single shell as opposed to individual Bash functions, calling things like `export KUBECONFIG` would set the environment variable into the calling script's shell context. Calling `export KUBECONFIG` from within a Bash function does not do this, at least not on Linux, which means that in order to kubectl to the KinD cluster now, I need to actually execute `export KUBECONFIG` after running `make kind-test`, which is annoying.

This patch modifies the `scripts/kind.sh:setup_kind_cluster` Bash function to create a `kubecontext` file inside the KinD cluster's build directory (`$ROOTDIR/build/clusters/$clustername/kubecontext`). This file contains exports for the `KUBECONFIG` environment variable, properly set to the KinD cluster's configuration. It also calls `kubectl config use-context` to set the current shell's kubectl context and use the `ack-system` namespace by default. Finally, the kubecontext file has a helper function calleld `controller_pod` that returns the Pod name for the controller under test.

This kubecontext file is shown in a block when we run `make kind-test` now:

```
2022-12-22T15:06:20-05:00 [INFO] Creating kubecontext file /home/jaypipes/go/src/github.com/aws-controllers-k8s/test-infra/scripts/../build/clusters/ack-test-7e1553e1/kubecontext ...

*******************************************
* To execute kubectl in the test context: *
*******************************************

source /home/jaypipes/go/src/github.com/aws-controllers-k8s/test-infra/scripts/../build/clusters/ack-test-7e1553e1/kubecontext
```

And sourcing the file in question sets the kubectl context appropriately and allows the developer to quickly access the controller pods and logs under test:

```
[jaypipes@thelio test-infra]$ source /home/jaypipes/go/src/github.com/aws-controllers-k8s/test-infra/scripts/../build/clusters/ack-test-7e1553e1/kubecontext
Switched to context "kind-ack-test-7e1553e1".
Context "kind-ack-test-7e1553e1" modified.
[jaypipes@thelio test-infra]$ k logs $(controller_pod) --tail 1
2022-12-22T20:10:55.733Z	DEBUG	ackrt	< r.deleteResource	{"account": "586748657421", "role": "", "region": "us-east-1", "kind": "Bucket", "namespace": "default", "name": "s3-bucket-vv97pk7q2imrxi", "generation": 16}
[jaypipes@thelio test-infra]$ k describe pods/$(controller_pod)
Name:         ack-s3-controller-dc744f569-74nz2
Namespace:    ack-system
Priority:     0
Node:         ack-test-7e1553e1-worker/172.18.0.3
Start Time:   Thu, 22 Dec 2022 15:07:13 -0500
Labels:       control-plane=controller
              pod-template-hash=dc744f569
Annotations:  <none>
Status:       Running
IP:           10.244.1.3
IPs:
  IP:           10.244.1.3
Controlled By:  ReplicaSet/ack-s3-controller-dc744f569
Containers:
  controller:
    Container ID:  containerd://e026ab4d9d9203e99fafba7702ad5067b011761272bae7784e7ea61fc49b0bd6
    Image:         aws-controllers-k8s:s3-v0.1.5-2-g2bbd673-dirty
    Image ID:      sha256:6a3e07cf23dac5a2755a1dd2ebfa57ccd29fe2147460205b666dc8604048b75a
    Port:          8080/TCP
    Host Port:     0/TCP
    Command:
      ./bin/controller
    Args:
      --aws-region
      $(AWS_REGION)
      --aws-endpoint-url
      $(AWS_ENDPOINT_URL)
      --enable-development-logging
      $(ACK_ENABLE_DEVELOPMENT_LOGGING)
      --log-level
      $(ACK_LOG_LEVEL)
      --resource-tags
      $(ACK_RESOURCE_TAGS)
      --watch-namespace
      $(ACK_WATCH_NAMESPACE)
    State:          Running
      Started:      Thu, 22 Dec 2022 15:07:14 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     100m
      memory:  300Mi
    Requests:
      cpu:     100m
      memory:  200Mi
    Environment:
      ACK_SYSTEM_NAMESPACE:            ack-system (v1:metadata.namespace)
      AWS_REGION:                      us-east-1
      AWS_ENDPOINT_URL:
      ACK_WATCH_NAMESPACE:
      ACK_ENABLE_DEVELOPMENT_LOGGING:  true
      ACK_LOG_LEVEL:                   debug
      ACK_RESOURCE_TAGS:               services.k8s.aws/controller-version=%CONTROLLER_SERVICE%-%CONTROLLER_VERSION%,services.k8s.aws/namespace=%K8S_NAMESPACE%
...
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-p7pn5 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  kube-api-access-p7pn5:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  25m   default-scheduler  Successfully assigned ack-system/ack-s3-controller-dc744f569-74nz2 to ack-test-7e1553e1-worker
  Normal  Pulled     25m   kubelet            Container image "aws-controllers-k8s:s3-v0.1.5-2-g2bbd673-dirty" already present on machine
  Normal  Created    25m   kubelet            Created container controller
  Normal  Started    25m   kubelet            Started container controller
```

Signed-off-by: Jay Pipes <jaypipes@gmail.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: scripts/kind.sh [hunk 1]
```diff
@@ -14,14 +14,34 @@ source "$SCRIPTS_DIR/lib/logging.sh"
...
+    info_msg "Creating kubecontext file $kubecontext_path ... "
+    cat <<EOF > "$kubecontext_path"
+export KUBECONFIG="$kubeconfig_path"
+export KUBECONTEXT="$kubecontext"
+kubectl config use-context "$kubecontext"
+kubectl config set-context --current --namespace=ack-system
 
+controller_pod() {
+  echo \$(kubectl get pods -ojsonpath='{.items[0].metadata.name}')
+}
```
Comment (A-Hilaly) [inline scripts/kind.sh:34]: How/Where is this function called? I don't see any mentions of it in test-infra repo

Comment (A-Hilaly) [reply scripts/kind.sh:34]: NVM i missed this
> Finally, the kubecontext file has a helper function calleld controller_pod that returns the Pod name for the controller under test.

File: scripts/kind.sh [hunk 2]
```diff
@@ -14,14 +14,34 @@ source "$SCRIPTS_DIR/lib/logging.sh"
...
-    export KUBECONFIG=$kubeconfig_path
+    info_msg "Creating kubecontext file $kubecontext_path ... "
+    cat <<EOF > "$kubecontext_path"
+export KUBECONFIG="$kubeconfig_path"
+export KUBECONTEXT="$kubecontext"
+kubectl config use-context "$kubecontext"
+kubectl config set-context --current --namespace=ack-system
 
+controller_pod() {
+  echo \$(kubectl get pods -ojsonpath='{.items[0].metadata.name}')
```
Comment (jljaco) [inline scripts/kind.sh:33]: How do we know that `.items[0]` represents the controller under test?  Do we need to be concerned about a user's kind cluster running multiple ACK controllers?

Comment (jaypipes) [reply scripts/kind.sh:33]: We don't :) This is just a best effort thing... that will work 99% of the time

Comment (RedbackThomson) [reply scripts/kind.sh:33]: Should this be querying the `ack-system` namespace? Or are we installing these into `default`?

**Note: I was going to suggest the following, but I realise we don't actually have the controller type within the pod label**
Could we at least filter by label to ensure we are getting the right *type* of controller pod?
```
kubectl get pods -o=jsonpath='{.items[?(@.metadata.labels.name=="ack-s3-controller")].metadata.name}'
```

Comment (jaypipes) [reply scripts/kind.sh:33]: > Should this be querying the `ack-system` namespace? Or are we installing these into `default`?

See line 30 above:

```diff
+ kubectl config set-context --current --namespace=ack-system
```

that sets the default namespace for the context to `ack-system` so ACK developers don't need to specify it.

> **Note: I was going to suggest the following, but I realise we don't actually have the controller type within the pod label** Could we at least filter by label to ensure we are getting the right _type_ of controller pod?
> 
> ```
> kubectl get pods -o=jsonpath='{.items[?(@.metadata.labels.name=="ack-s3-controller")].metadata.name}'
> ```

Yeah, I like the above idea...

**Pull request closed without merging**